src/map/achievement.c

```diff
@@ -0,0 +1,1693 @@
+
+	nullpo_retv(sd);
+	nullpo_retv(ad);
+
+	Assert_retv(progress > 0);
+	Assert_retv(obj_idx < VECTOR_LENGTH(ad->objective));
+
+	// find existing or create a new entry.
+	ach = achievement->find_or_create(sd, ad->id, true);
+
+	if (ach->completed_at)
+		return;
+	
+	// Check and increment the objective count.
+	if (!ach->objective[obj_idx] || ach->objective[obj_idx] < VECTOR_INDEX(ad->objective, obj_idx).goal) {
+		ach->objective[obj_idx] = min(progress + ach->objective[obj_idx], VECTOR_INDEX(ad->objective, obj_idx).goal);
+
+		// Check if the Achievement is complete.
+		if (achievement->check_complete(sd, ad)) {
+			achievement->validate_achieve(sd, ad->id);
+			ach->completed_at = time(NULL);
+		}
+
+		// update client.
+		clif->achievement_send_update(sd->fd, sd, ad);
+	}
+}
+
+/**
+ * Compare an absolute progress value against the goal of an objective.
+ * Does not add/increase progress.
+ * @param sd        pointer to map-server session data.
+ * @param ad        pointer to achievement data.
+ * @param obj_idx   index of the objective in question.
+ * @param progress  progress of the objective in question.
+ */
+void achievement_progress_set(struct map_session_data *sd, struct achievement_data *ad, const unsigned int obj_idx, const int progress)
+{
+	struct achievement *ach = NULL;
+
+	nullpo_retv(sd);
+	nullpo_retv(ad);
+
+	Assert_retv(progress > 0);
+	Assert_retv(obj_idx < VECTOR_LENGTH(ad->objective));
+
+	if (progress >= VECTOR_INDEX(ad->objective, obj_idx).goal) {
+		if ((ach = achievement->find_or_create(sd, ad->id, true)) && ach->completed_at)
+			return;
+
+		ach->objective[obj_idx] = VECTOR_INDEX(ad->objective, obj_idx).goal;
+
+		if (achievement->check_complete(sd, ad)) {
+			achievement->validate_achieve(sd, ad->id);
+			ach->completed_at = time(NULL);
+		}
+
+		clif->achievement_send_update(sd->fd, sd, ad);
+	}
+}
+
+/**
+* Checks if the given criteria satisfies the achievement's objective.
+* @param objective pointer to the achievement's objectives data.
+* @param criteria  pointer to the current session's criteria as a comparand.
+* @return true if all criteria are satisfied, else false.
+*/
+bool achievement_check_criteria(const struct achievement_objective *objective, const struct achievement_objective *criteria)
+{
+	bool equal = true;
+	int i = 0, j = 0;
+
+	nullpo_retr(false, objective);
+	nullpo_retr(false, criteria);
+
+	/* Monster Id */
+	equal = (objective->mobid && objective->mobid == criteria->mobid);
+
+	/* Item Id */
+	equal = (objective->unique.itemid && objective->unique.itemid == criteria->unique.itemid);
+
+	/* Weapon Level */
+	equal = (objective->unique.weapon_lv && objective->unique.weapon_lv == criteria->unique.weapon_lv);
+
+	/* Achievement Id */
+	equal = (objective->unique.achieve_id && objective->unique.achieve_id == criteria->unique.achieve_id);
+
+	/* Status Types */
+	equal = (objective->unique.status_type && objective->unique.status_type == criteria->unique.status_type);
```

Comment:
`unique` is an union. You need to know which field of the union has been used rather than probing them (reading the other fields will return garbage!)